### PR TITLE
[#10959] Paginate frontend call to webapi for large course

### PR DESCRIPTION
**Pull request by teikjun**
Part of #10959, follow-up to PR https://github.com/TEAMMATES/teammates/pull/10960

**Outline of Solution**

- Created component for `progress-bar`, added a service to handle the data flow
- Added `progress-bar` component as a child of `simple-modal` component
- Added a new type of `simple-modal`, `SimpleModalType.LOAD`, that displays the `progress-bar`

stuff
The test data used involved 600 students, 50 questions.

**Remarks**
- When the download size is small, the progress bar still loads smoothly so it doesn't look jittery - the progress bar just fills up very quickly. 
- I changed the color of Abort button from the primary color to danger color to indicate to the user that clicking the button is not the default action (unlike the usual `INFO` modal)
- I will pull the latest changes from upstream after review, before merging

### CODE REVIEW

File: src/web/app/components/loader-bar/loader-bar.component.scss [hunk 1]
```diff
@@ -7,7 +7,7 @@
 }
 
 /* stylelint-disable selector-pseudo-element-no-unknown */
-::ng-deep .progress {
+:host ::ng-deep .progress {
```
Comment (teikjun) [inline src/web/app/components/loader-bar/loader-bar.component.scss:10]: This change is not directly related to our issue but I thought I should fix it anyway.
Rationale: If the `::ng-deep` combinator is used without the `:host` pseudo-class selector, the style can bleed into other components. 
Source: https://angular.io/guide/component-styles

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts [hunk 2]
```diff
@@ -133,18 +134,19 @@ export class InstructorSessionResultPageComponent extends InstructorCommentsComp
...
-              private instructorService: InstructorService,
-              private route: ActivatedRoute,
-              private timezoneService: TimezoneService,
-              private simpleModalService: SimpleModalService,
-              private commentsToCommentTableModel: CommentsToCommentTableModelPipe,
-              statusMessageService: StatusMessageService,
-              commentService: FeedbackResponseCommentService,
-              commentToCommentRowModel: CommentToCommentRowModelPipe,
-              tableComparatorService: TableComparatorService) {
+    private feedbackQuestionsService: FeedbackQuestionsService,
```
Comment (daongochieu2810) [inline src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts]: inconsistent indentation?

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts [hunk 3]
```diff
@@ -169,37 +171,37 @@ export class InstructorSessionResultPageComponent extends InstructorCommentsComp
     }).subscribe((feedbackSession: FeedbackSession) => {
       this.session = feedbackSession;
       this.formattedSessionOpeningTime = this.timezoneService
-          .formatToString(this.session.submissionStartTimestamp, this.session.timeZone, TIME_FORMAT);
```
Comment (daongochieu2810) [inline src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts]: also indentation

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts [hunk 4]
```diff
@@ -468,15 +470,15 @@ export class InstructorSessionResultPageComponent extends InstructorCommentsComp
 
   downloadQuestionResultHandler(question: { questionNumber: number, questionId: string }): void {
     const filename: string =
-        `${this.session.courseId}_${this.session.feedbackSessionName}_question${question.questionNumber}.csv`;
+      `${this.session.courseId}_${this.session.feedbackSessionName}_question${question.questionNumber}.csv`;
 
     this.feedbackSessionsService.downloadSessionResults(
-        this.session.courseId,
```
Comment (daongochieu2810) [inline src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts]: same here

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts [hunk 5]
```diff
@@ -421,25 +422,25 @@ export class InstructorSessionResultPageComponent extends InstructorCommentsComp
...
         this.feedbackSessionsService.downloadSessionResults(
-            this.session.courseId,
-            this.session.feedbackSessionName,
-            Intent.INSTRUCTOR_RESULT,
-            this.indicateMissingResponses,
-            this.showStatistics,
-            this.questionsModel[k].question.feedbackQuestionId,
-            this.section.length === 0 ? undefined : this.section,
-            this.section.length === 0 ? undefined : this.sectionType,
+          this.session.courseId,
```
Comment (daongochieu2810) [inline src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts]: same here

**Review by daongochieu2810 — CHANGES_REQUESTED** [review]
I suggest we show the number of responses downloaded e.g ```53 responses downloaded...``` instead of ```53%``` just to make it clearer for instructors

**Comment by teikjun** [comment]
Thanks for the review, will fix the indentation issues! 

The download is split by question level, so the suggested change would either look like "25/50 questions downloaded..." or "15000/30000 responses downloaded...". I think that revealing this implementation detail can confuse the user. 

I think it is better to just show users the download progress as a percentage, as they are only concerned with "How long more until download is done?".

**Comment by daongochieu2810** [comment]
> Thanks for the review, will fix the indentation issues!
> 
> The download is split by question level, so the suggested change would either look like "25/50 questions downloaded..." or "15000/30000 responses downloaded...". I think that revealing this implementation detail can confuse the user.
> 
> I think it is better to just show users the download progress as a percentage, as they are only concerned with "How long more until download is done?".

For large numbers, we can shorten them to smth like 1k 2k. Tbh I'm not sure if instructors would prefer being shown the number of responses/questions downloaded or just the progress. Personally I would want to know the exact number

File: src/web/app/components/loader-bar/loader-bar.component.scss (hunk 1 above)
Comment (moziliar) [reply src/web/app/components/loader-bar/loader-bar.component.scss:10]: Any visual of this? Although can't comment much on scss.

**Review by moziliar — CHANGES_REQUESTED** [review]
Very impressed by the look. Really decent UX to me.
Do fix all the formatting issue (else CI complains anyways) with `npm run lint`.

**Comment by moziliar** [comment]
@daongochieu2810 can we table this to the actual instructor user to ask about the opinion?

**Comment by damithc** [comment]
I'm OK with the look. Don't mind leaving it as percentages. Using raw numbers will lead to more questions such as which entity type to use for the counter.
Unless the operations results in a file being saved locally, perhaps use 'loading' or 'fetching' rather than 'downloading'?

File: src/web/app/components/loader-bar/loader-bar.component.scss (hunk 1 above)
Comment (teikjun) [reply src/web/app/components/loader-bar/loader-bar.component.scss:10]: `:host` scopes the styles to this component, so that it does not affect other components. The styles for this component is unchanged! :)

**Comment by moziliar** [comment]
It seems that you have caused some other formatting issues in the java files. Please fix those. @teikjun

**Comment by moziliar** [comment]
Can you reset the last commit and force push to this branch? I'm not sure if your own branch should merge master. You can tell from the 9k LoC. I've just updated my own branch.

**Comment by teikjun** [comment]
Done! Feel free to merge if it looks fine, I'll make the remaining changes (incl. the modal overlay and enabling batch processing for the button on Session/Home tab) in a separate PR.

_(Question below is not directly related to this PR)_
May I ask how do you guys configure your code formatter for typescript files? For instance, do you use the default IDE formatter, do some manual configuration in the settings, or just format the code manually?

The linter accepts multiple ways of code formatting, so it doesn't catch some formatting issues, e.g. using continuation indent of 2 instead of 4 (like in the initial version of this PR). I noticed that some code in the main branch has similar formatting issues as well. For example:

In `instructor-session-base-page.component.ts` , 2 space continuation indent:

In the same file, 4 space continuation indent:

**Comment by moziliar** [comment]
I'd say whenever I find some formatting from the IDE problematic, I'd just change it right away. Always run `npm run lint` before pushing, as failing lint either waste your own time on waiting for CI to lint or other ppl to review (in this case, the lint issue blocks for almost a week).
About formatting, just follow npm lint. We pointed out about your formatting issue as they seem to be formatted differently from the target branch for no reason, and therefore likely incurring a CI lint failure.

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts [hunk 6]
```diff
@@ -287,18 +289,18 @@ export class InstructorSessionResultPageComponent extends InstructorCommentsComp
...
+            const responses: QuestionOutput = resp.questions[0];
+            this.questionsModel[questionId].responses = responses.allResponses;
+            this.questionsModel[questionId].statistics = responses.questionStatistics;
+            this.questionsModel[questionId].hasPopulated = true;
+
+            this.preprocessComments(responses.allResponses);
+          }
+        }, (resp: ErrorMessageOutput) => {
+          this.statusMessageService.showErrorToast(resp.error.message);
+        });
```
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts]: various parts of your codes have the spacing changed. I can't merge as merging it is very likely to incur CI failure in my PR, which would demand a commit to fix (either me or you). Please do it here.

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts [hunk 7]
```diff
@@ -366,12 +368,12 @@ export class InstructorSessionResultPageComponent extends InstructorCommentsComp
       const modalContent: string = `An email will be sent to students to inform them that the session has been unpublished and the session responses
           will no longer be viewable by students.`;
       modalRef = this.simpleModalService.openConfirmationModal(
-          `Unpublish this session <strong>${ this.session.feedbackSessionName }</strong>?`,
+          `Unpublish this session <strong>${this.session.feedbackSessionName}</strong>?`,
```
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:371]: Leave spacing for interpolated string. It seems more just that way.

File: src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts [hunk 8]
```diff
@@ -366,12 +368,12 @@ export class InstructorSessionResultPageComponent extends InstructorCommentsComp
...
           SimpleModalType.WARNING, modalContent);
     } else {
       const modalContent: string = 'An email will be sent to students to inform them that the responses are ready for viewing.';
       modalRef = this.simpleModalService.openConfirmationModal(
-          `Publish this session <strong>${ this.session.feedbackSessionName }</strong>?`,
+          `Publish this session <strong>${this.session.feedbackSessionName}</strong>?`,
```
Comment (moziliar) [inline src/web/app/pages-instructor/instructor-session-result-page/instructor-session-result-page.component.ts:376]: ditto.

**Review by moziliar — CHANGES_REQUESTED** [review]
Do fix changes that are totally unnecessary, of which almost all stemmed from formatting issue. Do not change the formatting unless it is a PR explicitly doing so.

**Review by moziliar — APPROVED** [review]
LGTM!

**Review by daongochieu2810 — APPROVED** [review]
lgtm
